stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -793,6 +793,57 @@ function dot(x::AbstractArray, y::AbstractArray)
     s
 end
 
+"""
+    dot(x, A, y)
+
+Compute the generalized dot product `x' * A * y` between two vectors `x` and `y`.
+For complex vectors, the first vector is conjugated.
+"""
+dot(x::Number, A::Number, y::Number) = conj(x) * A * y
+
+function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
+    m, n = size(A)
+    (length(x) == m && n == length(y)) || throw(DimensionMismatch())
+    if iszero(m) || iszero(n)
+        return dot(zero(eltype(x)), zero(eltype(A)), zero(eltype(y)))
+    end
+
+    if m == n
```

Comment:
These test the numeric values in the matrix A at runtime for various properties, if these are not encoded in the type. If returning true, then we have already iterated over the entire matrix, and hence already paid large parts of the cost of this function (i.e. memory transfers).

Such runtime tests for useful special properties make sense for expensive operations (e.g. matrix factorizations) where the number of floating point ops is much larger than the number of memory transfers. I don't think they make sense here.

Reply:
I added them to trigger comments like these, either in favor or against. I knew there is some effort related to these checks, but wasn't sure how they would weigh against the benefits from using the specialized methods.

Reply:
I agree that the checks for specialized matrix types should be omitted — in this function they are as expensive as the dot product itself.

Reply:
This is removed now and forever.